docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+  // running, this field will not be present.
+  google.protobuf.Timestamp endTime = 7;
+  // The job's original spec.
+  JobSpec spec = 8;
+}
+
+message CommandSpec {
+  // The command name to run. Required.
+  string command = 1;
+  // The command's arguments, not containing the command name itself.
+  repeated string args = 2;
+  // Optional additional environment variables to set for the command.
+  // These will be merged with the job server's environment variables.
+  repeated string env = 3;
+  // The working directory for the command. If not specified, the job server's
+  // working directory will be used.
+  optional string cwd = 4;
+  // Optional user id to run the command as.
+  // If not specified, the job server's user id will be used.
+  optional int32 uid = 5;
+  // Optional group id to run the command as.
+  // If not specified, the job server's group id will be used.
+  optional int32 gid = 6;
+  // Optional data to be written to the process's stdin pipe when it starts.
+  // If not specified, the process's stdin will be set to /dev/null.
+  optional bytes stdin = 7;
+}
+
+message ProcessOutput {
+  // A chunk of output from the process's combined stdout and stderr streams.
+  //
+  // This is a raw byte stream, and may contain arbitrary binary data,
+  // depending on the command being run. It is up to the client to interpret
+  // the output correctly based on the command being run. No guarantees are
+  // made about the size of each chunk, or the frequency at which they
+  // are sent.
+  bytes output = 1;
+}
+
+message ResourceLimits {
+  // Process CPU limit defined in milli-cores (1000 = 1 core/vcpu)
+  optional int64 cpu = 1;
+  // Process memory limit defined in bytes
+  optional MemoryLimits memory = 2;
+  // Process IO limits for devices
+  optional IODeviceLimits io = 3;
+}
+
+message MemoryLimits {
+  // Memory soft limit defined in bytes. Processes exceeding the limit are
+  // not OOM-killed, but may be throttled by the kernel.
+  optional int64 softLimit = 1;
+  // Memory limit defined in bytes. If exceeded, the process will be OOM-killed.
+  optional int64 limit = 2;
+}
+
+message IODeviceLimits {
+  // Limits for individual devices
+  repeated DeviceLimit devices = 1;
+}
+
+message DeviceLimit {
+  // Device id (major:minor)
+  string deviceId = 1;
+  // Limits for the device
+  IOLimits limits = 2;
+}
+
+message IOLimits {
+  // Limit for read operations in bytes per second
+  optional int64 readBps = 2;
+  // Limit for write operations in bytes per second
+  optional int64 writeBps = 3;
+  // Limit for read operations in IOPS
+  optional int64 readIops = 4;
+  // Limit for write operations in IOPS
+  optional int64 writeIops = 5;
+}
+
+message OutputRequest {
+  JobId id     = 1;
+  bool  follow = 2;
+}
+
+```
+
+Authorization uses a very simple RBAC system. The RBAC configuration is defined as follows:
+
+```protobuf
+syntax = "proto3";
```

Comment:
Do these need to be protobuf objects?

Reply:
They don't, but (hypothetically, if this were a real project) defining these types as protos would make it much easier to create an admin API to allow users to update these rules on the fly, e.g.

```protobuf
service RBAC {
  rpc GetConfig(google.protobuf.Empty) returns (rbac.v1.Config);
  rpc UpdateConfig(rbac.v1.Config) returns (google.protobuf.Empty);
  rpc ListRoles(google.protobuf.Empty) returns (RoleList);
  // etc...
}